implement module.registerHooks() to run synchronous module customization hooks in thread

joyeecheung opened the pull request:
This implements part of the proposal in https://github.com/nodejs/loaders/pull/198 - for the motivations, see https://github.com/nodejs/node/issues/52219 - this fills in the gap where being able to run in-thread and support `require()` is more important for a hook than being able to run asynchronous code (even then, packages like https://www.npmjs.com/package/everysync allows user to sync-ify async code off-thread themselves), especially for a large amount of CJS loader moneky-patchers out there to migrate to an officially supported API easily.

<img width="999" alt="Screenshot 2024-11-19 at 16 01 47" src="https://github.com/user-attachments/assets/a8b04978-f3f6-4b6d-adfb-63710ad8ddf1">

The synchronous, in-thread hooks are also a lot easier to support universally in the loaders and reduce the number of dark corners where the hooks cannot run or behave in a surprising manner - I guess the amount of "this caveat of asynchronous hooks does not apply to the synchronous hooks" added in the documentation kind of also already proves the point...

For now only `resolve` and `load` are implemented to reach parity with `module.register()`. I left the exports hook to https://github.com/joyeecheung/node/tree/export-hooks which will be a follow-up when we decide the timing at which the post-load hook for CJS should run.

The commit `module: use synchronous hooks for preparsing in import(cjs)` isn't strictly necessary and could split into a follow-up instead, but I think it's nice to have it here instead of leaving out that corner uncovered.
 
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

joyeecheung commented:
Added a lot of tests. This should now be ready for review.

guybedford commented on doc/api/module.md:
```diff
@@ -714,6 +740,35 @@ export async function load(url, context, nextLoad) {
+may be terminated by the main thread at any time, so do not depend on
+asynchronous operations (like `console.log`) to complete.
+
+#### Synchronous hooks accepted by `module.registerHooks()`
```
Can we make the first section here something like the following:

```
#### Registering synchronous hooks

Unlike asynchronous hooks, synchronous hooks only run on the current thread, and are registered on that thread.

It is recommended that registrations made by the `module.registerHooks()` API define themselves in a separately defined hooks module that can be imported by all threads of the following form:

... demonstrate the top-level registration and --import model, per below...
```

I think that giving recommendation and guidance for hooks library authors to the right patterns here is important to ensure cross-thread semantics as we've discussed before.

joyeecheung replied:
Added a few sentence about this. I would prefer to just leave more elaborated explanations outside of an...API documentation.

guybedford commented on doc/api/module.md:
````diff
@@ -1170,6 +1281,57 @@ async function getPackageType(url) {
+registerHooks({ load });
+```
+
+#### Running hooks
````
I'm basically saying we should move this section first before going into the hooks API details.

joyeecheung replied:
I think this should either be done in a separate PR because the whole section predated this PR, or just moved out of the API documentation and into some tutorial entirely. Making structural changes in a semver-minor PR is a great way to make backporting a nightmare.

guybedford reviewed (commented):
We've discussed this in the past and my previous feedback still stands that overall I'm very much for this direction, with the caveat that we just need to ensure we guide hooks authors to patterns that work well for ensuring consistent behaviours including for environments that use workers.

Specifically it would be great if we can recommend the `--import` top-level hook registration pattern quite strongly, as that naturally works with workers and it's worth calling out that this does that explicitly when documenting `registerHooks`.

I do have one feature request I'd love to see here though - a one-way function to disable hooks for security `module.lockHooks()`, with no arguments, that disallows further `register` or `registerHooks` calls.

Something to this effect would be a huge help to ensure runtime hooks registered on arbitrary function calls is not a thing, and we do think about the execution model as "register hooks" moving into a stage of "hooks are registered and can't be registered", rather than thinking of hooks as being able to happen at any time.

joyeecheung commented:
> I do have one feature request I'd love to see here though - a one-way function to disable hooks for security module.lockHooks(), with no arguments, that disallows further register or registerHooks calls.

That seems out of scope of this PR? If it was fine not having it for `module.register`, why does it need to happen in this PR?

guybedford commented:
Having a hook lockdown phase is not a requirement for this PR by any means, but would be a nice to have to clarify the early hook registration model sooner rather than later.

guybedford commented:
Right, so it's our job to educate hook authors how to write hooks to support multi threading (or not) and how to avoid recursive application etc. etc.

With regards to hook "lockdown" I think this is something that the application owner would care about - how do I know that an arbitrary npm package import isn't going to be doing hooks without me knowing about it? How do I know I'm not depending on a package that's automatically adding a "jsx" hook and changing the way my local code runs? A lockdown feature is for the application code owner's benefit. Ideally lockdown by default would be great too to ensure we don't enter a Node.js future where hooking the loader is common for library dependencies.

joyeecheung commented:
> Right, so it's our job to educate hook authors how to write hooks to support multi threading (or not) and how to avoid recursive application etc. etc.

While I agree better documentation should exist, I think we shouldn't really dive into too much details in the API documentation. It should either be in the learn section in the website, or just have dedicated repository for tutorials + examples like https://github.com/nodejs/node-addon-examples and the new https://github.com/nodejs/package-examples that we are spinning (I also left comment in this PR that I think the example loaders should be moved in its own repo, and refrained from adding too many complementary examples for the new API, I also noticed that some existing examples in the documentation didn't even work, but it felt out of scope to fix them here).

guybedford commented:
@joyeecheung to give an example, I was not aware `--import` applied for all threads and this came as a surprise to me personally. This information is valuable.

joyeecheung commented:
Yes, though not all valuable information belongs in the API documentation ;) One can argue that https://nodejs.github.io/node-addon-examples/ is full of valuable information, but it's better to have a dedicated tutorial on a complex topic like this, instead of trying to elaborate on it in the API documentation. I think we can make do with one or two sentence about `--import` for this PR, and leave detailed explanation as a TODO, and don't pile even more information that should be better hosted elsewhere into the API documentation.

GeoffreyBooth commented on lib/internal/modules/sync_hooks.js:
```diff
@@ -0,0 +1,338 @@
+let nextHookId = 0;
+
+class ModuleHooks {
+  constructor(resolve, load) {
```
Please add JSDoc.

GeoffreyBooth commented on lib/internal/modules/sync_hooks.js:
```diff
@@ -0,0 +1,338 @@
+};
+
+// TODO(joyeecheung): taken an optional description?
+function registerHooks(hooks) {
```
Please add JSDoc.

GeoffreyBooth commented on lib/internal/modules/sync_hooks.js:
```diff
@@ -0,0 +1,338 @@
+};
+
+class ModuleLoadContext {
+  constructor(format, importAttributes, conditions) {
```
Please add JSDoc.

GeoffreyBooth commented on lib/internal/modules/sync_hooks.js:
```diff
@@ -0,0 +1,338 @@
+  }
+};
+
+/**
+ * Load module source for a url, through a chain hooks if they exist.
+ * @param {string} url
+ * @param {string|undefined} originalFormat
+ * @param {ImportAttributes|undefined} importAttributes
+ * @param {string[]} conditions
+ * @param {(url: string, context: ModuleLoadContext) => ModuleLoadResult} defaultLoad
+ * @returns {ModuleLoadResult}
+ */
+
+let decoder;
+function loadWithHooks(url, originalFormat, importAttributes, conditions, defaultLoad) {
```
```suggestion
let decoder;
/**
 * Load module source for a url, through a hooks chain if it exists.
 * @param {string} url
 * @param {string|undefined} originalFormat
 * @param {ImportAttributes|undefined} importAttributes
 * @param {string[]} conditions
 * @param {(url: string, context: ModuleLoadContext) => ModuleLoadResult} defaultLoad
 * @returns {ModuleLoadResult}
 */
function loadWithHooks(url, originalFormat, importAttributes, conditions, defaultLoad) {
```

GeoffreyBooth commented on lib/internal/modules/sync_hooks.js:
```diff
@@ -0,0 +1,338 @@
+}
+
+/**
+ * Resolve module request to a url, through a chain hooks if they exist.
```
```suggestion
 * Resolve module request to a url, through a hooks chain if it exists.
```

GeoffreyBooth commented on test/fixtures/module-hooks/log-user.cts:
```diff
@@ -0,0 +1,4 @@
+
```
```suggestion
```

GeoffreyBooth commented on test/fixtures/module-hooks/log-user.mts:
```diff
@@ -0,0 +1,5 @@
+
```
```suggestion
```

GeoffreyBooth commented on test/parallel/test-bootstrap-modules.js:
```diff
@@ -98,6 +98,7 @@ expected.beforePreExec = new Set([
   'Internal Binding contextify',
   'NativeModule internal/vm',
   'NativeModule internal/modules/helpers',
+  'NativeModule internal/modules/sync_hooks',
```
Perhaps just call this `hooks`, since we plan to remove the async hooks?

```suggestion
  'NativeModule internal/modules/hooks',
```

joyeecheung replied:
Because there's `internal/modules/esm/hooks`, and I think it's even more confusing to have two files that are named `hooks`, but then it doesn't make sense to add the new hooks there since this is not esm-only. Then again, I don't want to move that file to `internal/modules` in this PR, because that would make backporting a lot harder.

GeoffreyBooth replied:
How about `customization_hooks` then? It won’t make sense to call this `sync_hooks` once the async hooks are gone.

joyeecheung replied:
`customization_hooks` SGTM (and better, because hooks can mean too many things..)

GeoffreyBooth commented on test/parallel/test-module-hooks-import-wasm.mjs:
```diff
@@ -0,0 +1,35 @@
+// Flags: --no-experimental-wasm-modules
```
Please put all the new tests under `test/es-module/`. It’s been our practice for years to run the modules tests as a subset via `tools/test.py --mode=release test/es-module`. At this point perhaps `es-module` should just be named `modules`, but it’s very useful to be able to run all the modules-related tests in a few seconds without needing to run the full `test/parallel` set of tests.

joyeecheung replied:
I don't think they should be put under `test/es-module`, because they are not ESM specific. Though I could add them to `test/modules` (not what I would recommend, because then you would start moving `test/es-modules` to `test/modules` and make backporting harder), or `test/module-hooks` (less impactful so would be okay-ish for backporting).

> It’s been our practice for years to run the modules tests as a subset via tools/test.py --mode=release test/es-module

I don't think that's the right way to "run all tests related to modules" in the first place, because you would be missing a lot of CJS tests with this. Perhaps you meant it's a practice to run "tests that are only related to ESM"? Then that doesn't apply to the tests added here, because they are not ESM-only.

> but it’s very useful to be able to run all the modules-related tests in a few seconds without needing to run the full test/parallel set of tests.

In case you didn't know, you can run `tools/test.py "test/parallel/test-module-hook*"`. If you want a shortcut to run all module-related tests, I would suggest to add a Makefile target that runs `tools/test.py es-module`, `tools/test.py test/parallel/*module*` which would provide proper coverage, instead of counting on `tools/test.py es-module` to cover all the module tests, since it already doesn't do that.

GeoffreyBooth replied:
I think in general we shouldn’t be dumping more and more tests into `test/parallel`. It’s already unusable in a file explorer, having thousands of files within it. If you don’t want to put them in `test/es-module` then perhaps this can be the PR where we create `test/module`, and a follow-up can migrate `test/es-module` into it?

joyeecheung replied:
> a follow-up can migrate test/es-module into it?

I feel that's going to create a lot of headaches for backports (which I absolutely won't enjoy when I backport `require(esm)` to v20..). I would prefer `test/module-hooks` for a smaller scope.

GeoffreyBooth replied:
> I would prefer `test/module-hooks` for a smaller scope.

I was thinking maybe that `test/*` could mirror `lib/*` and `src/*`, where we have `lib/modules`, but I guess it can't hurt to be even more specific.

joyeecheung replied:
I feel that there are many mistakes made in the `lib/something` and `src/something` naming choices (some by myself which I know regret about as they just make the file harder to find for me, as the category may not really make too much sense)

GeoffreyBooth commented on test/parallel/test-module-hooks-preload.js:
```diff
@@ -0,0 +1,49 @@
+const fixtures = require('../common/fixtures.js');
+const { spawnSyncAndAssert } = require('../common/child_process.js');
+
+spawnSyncAndAssert(process.execPath,
```
Why do these all need to be sync and run serially? Can’t we use `spawnPromisified` to run them all in parallel?

joyeecheung replied:
`spawnPromisified` provides very poor output when the test fails, `spawnSyncAndAssert` provides a lot more details e.g. when it fails it logs the commands and the environment variables, and both the stderr and stdout. If someone ports an async alternative that provides as much information as `spawnSyncAndAssert` I could consider using them, until then, I find `spawnPromisified` to be very frustrating when the test fails.

For child_process.spawn on a very fast process, you don't necessarily get a lot out of parallelism, considering the overhead of async handling and output buffering.

GeoffreyBooth commented on test/parallel/test-repl.js:
```diff
@@ -599,7 +599,9 @@ const errorTests = [
-      /^ {4}at .*/, // at Module._load
+      /^ {4}at .*/, // At defaultResolveImpl
+      /^ {4}at .*/, // At resolveForCJSWithHooks
+      /^ {4}at .*/, // at Function._load
```
Or we can update the `send` line a few lines up to be like `send: 'Error.stackTraceLimit = 1; require("internal/repl")',` and then just delete all the lines after 601 and then we don’t need to keep updating this test as the call stack within the CommonJS loader changes.

GeoffreyBooth reviewed (approved):
Great work! I’m excited to see this land.

joyeecheung commented on doc/api/module.md:
````diff
@@ -693,8 +703,24 @@ register('./my-hooks.mjs', {
 });
 ```
 
+### In-thread Synchronous Customization Hooks
````
> I wouldn’t want someone seeing register as more mature and more stable than registerHooks and therefore a safer choice.

I think that is an accurate description for now, considering `module.registerHooks()` is not yet even released, and `module.register()` has already been used by several projects. The future migration path should be a follow up after we actually get some feedback from the user land. This is a semver-minor PR and we should keep it so. The problem of the off-thread hooks came from dictating a path before sufficient real-world feedback was collected, and I don't think we should make the same mistake again just because we believe "this time, we got it right!" (I hope we are, but this isn't up to us).

GeoffreyBooth replied:
If you don’t want to include anything about “we consider this to be the way forward” just yet, fine, but I still think that the new hooks should go first; and the explanation about how the hooks compare with each other should live with the async hooks rather than with the sync hooks. That way when we remove the async hooks, the comparison text is already there and goes away cleanly.

joyeecheung replied:
I spent some time to move it and it took way to much effort to do an overhaul of the texts to describe the sync hooks first, because the existing texts always assume they are talking about the async ones and need wording changes everywhere. I don't think this needs to be done in the initial PR that's mostly just implementing things. Can you open a follow-up PR to do the rewrite?

joyeecheung commented on doc/api/module.md:
```diff
@@ -693,8 +703,24 @@ register('./my-hooks.mjs', {
+
+> Stability: 1.1 - Active development
+
+`module.registerHooks()` is a lighter weight alternative to `module.register()`
+which takes hook functions directly and run them in the same thread where the
+modules are loaded. The hook functions are synchronous, which is necessary to
```
I think for people who are already using `module.register()` and trying to figure out what the new API offers, this should be noted. We can remove it when `module.register()` actually gets removed, which would take a while because the usage out there already warrants a deprecation cycle, and in that case this would also serve as what all deprecation needs in the documentation - how to migrate to the alternative of the deprecated API.

joyeecheung commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -1858,3 +2040,4 @@ ObjectDefineProperty(Module.prototype, 'constructor', {
 
 // Backwards compatibility
 Module.Module = Module;
+Module.registerHooks = registerHooks;
```
Because `require('module') === Module`, `module.register` is also added to `Module`

https://github.com/nodejs/node/blob/e312d60e3f81dfa26d5922f3c23a371b2b2ce343/lib/module.js#L19

it's just a matter of adding it in `modules/cjs/loader.js` directly, or importing it and adding it and re-exporting it in `module.js`.

https://github.com/nodejs/node/blob/e312d60e3f81dfa26d5922f3c23a371b2b2ce343/lib/module.js#L28

jasnell commented on lib/internal/modules/sync_hooks.js:
```diff
@@ -0,0 +1,337 @@
+  // TODO(joyeecheung): we may want methods that allow disabling/enabling temporarily
+  // which just sets the item in the array to undefined temporarily.
+  // TODO(joyeecheung): this can be the [Symbol.dispose] implementation to pair with
+  // `using` when the explicit resource management proposal is shipped by V8.
```
May as well add the `Symbol.dispose` implementation now to prepare since we have it available.

legendecas replied:
Agreed this could be a follow-up. TBH, I don't really find how this could be helpful in the reality.

joyeecheung replied:
Yeah someone suggested this in the proposal PR IIRC, though I think using this with `using ` would be a rare choice - if you want something temporarily customized in the same file, chances are you are going to anchor it to other smaller parts of a bigger hook that just stays in the chain?

legendecas commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -1015,6 +1029,153 @@ function getExportsForCircularRequire(module) {
+function loadBuiltinWithHooks(id, url, format) {
+  if (loadHooks.length) {
+    url ??= `node:${id}`;
+    // TODO(joyeecheung): do we really want to invoke the load hook for the builtins?
```
Maybe the result could be asserted that `loadResult.source === null` if the format is not overridden, so that the result is not silently discarded?

legendecas replied:
Sounds good to me

guybedford commented on doc/api/module.md:
```diff
@@ -513,52 +531,89 @@ changes:
+
+There are two types of module customization hooks that are currently supported:
+
+1. `module.register(specifier[, parentURL][, options])` which takes a module that
+   exports asynchronous hook functions. The functions are run on a separate loader
+   thread.
+2. `module.registerHooks(options)` which takes synchronous hook functions that are
+   run directly on the thread where the module is loaded.
```
Why don't we just put the sync hooks first?

guybedford commented on doc/api/module.md:
```diff
@@ -513,52 +531,89 @@ changes:
-using the `--import` flag:
+Module resolution and loading can be customized by:
+
+1. Registering a file which exports a set of asynchronous hook functions, using the
+   [`register`][] method from `node:module`,
+2. Registering a set of synchronous hook functions using the [`registerHooks`][] method
+   from `node:module`.
```
Again, might be worth putting sync hooks first as the recommended path.

guybedford commented on doc/api/module.md:
````diff
@@ -513,52 +531,89 @@ changes:
+   from `node:module`.
+
+The hooks can be registered before the application code is run by using the
+[`--import`][] or [`--require`][] flag:
 
 ```bash
 node --import ./register-hooks.js ./my-app.js
+node --require ./register-hooks.js ./my-app.js
 ```
 
 ```mjs
 // register-hooks.js
+// This file can only be require()-ed if it doesn't contain top-level await.
+// To register asynchronous, off-thread hooks, use module.register().
 import { register } from 'node:module';
-
 register('./hooks.mjs', import.meta.url);
 ```
````
Again, for this first example, it would be great to see the sync hooks workflow first, then have the alternative workflow for async hooks provided.

guybedford commented on doc/api/module.md:
```diff
@@ -513,52 +531,89 @@ changes:
-files are imported, including the entry point of the application. Alternatively,
-`register` can be called from the entry point, but dynamic `import()` must be
-used for any code that should be run after the hooks are registered:
+Using `--import` or `--require` ensures that the hooks are registered before any
+application files are imported, including the entry point of the application.
```
```suggestion
Using `--import` or `--require` ensures that the hooks are registered before any
application files are imported, including the entry point of the application and for
any worker threads by default as well.
```

guybedford reviewed (commented):
The readme context is looking a lot better, my main remaining suggestion would be to state the sync hooks first as the recommended path.

Furthermore, as far as I'm aware there is rough consensus to deprecate the async customization hooks. Are there plans to follow-up with a deprecation PR at all?

GeoffreyBooth commented on lib/internal/modules/sync_hooks.js:
```diff
@@ -0,0 +1,338 @@
+  // which just sets the item in the array to undefined temporarily.
+  // TODO(joyeecheung): this can be the [Symbol.dispose] implementation to pair with
+  // `using` when the explicit resource management proposal is shipped by V8.
+  deregister() {
```
JSDoc isn’t just about types, it’s also to explain the purpose of the function. Deregister what? A particular hook? All hooks? It’s not obvious from the method signature, since the method takes no input.

In general though we decided a while back to require JSDoc comments for all functions and methods in the modules codebase, enforced via lint rule, so I’m also making a suggestion that I would assume should help the lint check pass.

joyeecheung replied:
I think it being an instance method is already self-explanatory and I find redundant comments making the code harder to read (one of the problems I find with the module loader code is that a lot of comments are only there for comment's sake without providing any additional information that can't already be understood from naming). But sure I can add a sentence explaining it deregister the instance.

joyeecheung commented:
I tried moving the docs of sync hooks first, but it turned out to be huge amount of work, because most of the docs shared for both starts with an example of the async one and the text & example now all have to be rewritten :( The doc overhaul is way too much work for a initial semver-minor PR that's just trying to add a variant that's less broken...

JakobJingleheimer commented on doc/api/module.md:
```diff
@@ -581,18 +636,52 @@ register('http-to-https', pathToFileURL(__filename));
+
+// The synchronous hooks affect import, require() and user require() function
+// created through createRequire().
+import('./my-app.js');
```
```suggestion
await import('./my-app.js');
```

GeoffreyBooth replied:
This is a CommonJS code block, it can’t get top-level `await`; but it can get un-awaited `import()`.

joyeecheung commented:
I think I've addressed most of the comments so far, except the one about changing the documentation to make sync hooks go first, because I think that would lead to an overhaul of the documentation as most of the explanatory texts are taking about async hooks examples and this PR mostly just append something saying what's different in sync hooks and that is not a huge amount of changes. But actually talking about sync hooks first means all of the texts and examples need to be reworded otherwise it looks incorrect logically. Also this is just an API that's not yet released and I feel that we could do the reordering when we have it more battle tested.

Other than the part about doc ordering I think this is good to go - incidentally in the past 24 hours I've already seem two glitches ([1](https://github.com/nodejs/node/issues/56127), [2](https://github.com/nodejs/node/pull/56040#issuecomment-2515172973)) coming from incorrect CJS loader monkey patching after require(esm) is unflagged on v22. So I hope we can get it over with sooner than later and stop the spread of incorrect CJS loader monkey patching in the ecosystem.

PTAL again, thanks @legendecas, @JakobJingleheimer, @guybedford, @GeoffreyBooth, @jasnell

legendecas reviewed (approved):
I tried to apply the sync loader hooks inside the ShadowRealms and it works really well without worker thread overheads. Looking forward to having this initial work landed and getting this improved.

joyeecheung commented:
Rebased to resolve conflicts in module.md (this is the second time now, kinda prove my point about what backporting nightmare it can be if the doc has to be overhauled in this PR..)

Can you take a look again? Thanks @legendecas, @JakobJingleheimer, @guybedford, @GeoffreyBooth, @jasnell

guybedford reviewed (approved):
Happy to leave the sync reordering to a deprecation PR for the async variant in due course.

joyeecheung commented:
Note to releasers: please use the OP text (modulo background) for summary in the change log. The graph can be left out I suppose.

This depends on at least two refactoring PRs that landed before: https://github.com/nodejs/node/pull/54769 and https://github.com/nodejs/node/pull/55590 .

ruyadorno commented:
This commit does not land cleanly on `v22.x-staging` and will need manual backport in case we want it in **v22.x**.

joyeecheung commented:
I think it's worth backporting to v22 (since the whole point is to get more people off CJS monkey patching), though I am not 100% sure the refactoring here plays well with er, existing CJS monkey patching (I tried my best to make it play, but one never knows how crazy the patching can be). So we should wait and see how it goes in 23 for a bit before backporting.

ThiH69 reviewed (commented):
~~[]()~~

joyeecheung commented:
This needs to be backported together with https://github.com/nodejs/node/pull/56402

goloveychuk commented:
@joyeecheung hey, I see that `require.resolve` is calling `Module._resolveFilename`,
which does not have `resolveWithHooks` calling. Even the opposite, `resolveForCJSWithHooks` is calling `Module._resolveFilename`.
Which means sync hooks will not intercept `require.resolve`, although docs implies it should work. 
It seems like `resolveWithHooks` logic should be implemented internally in Module._resolveFilename.
